src/DisputeHandler.ts

```diff
@@ -55,164 +72,281 @@ class DisputeHandler {
-        return this.createDispute(
-            block.transaction.header.forkCnt,
-            NO_PARTICIPANT_TO_FOLD,
-            INITIAL_TRANSACTION_COUNT,
-            [proof]
-        );
+    private collectStateProof(forkCnt: number): StateProofStruct {
+        const forkProof =
+            this.agreementManager.forkService.getForkProof(forkCnt)!;
+        const signedBlocks =
+            this.agreementManager.getForkProofSignedBlocks(forkCnt);
+        const stateProof: StateProofStruct = {
+            forkProof: forkProof,
+            signedBlocks: signedBlocks
+        };
+        return stateProof;
     }
 
-    public onDispute(dispute: DisputeStruct): Promise<void> {
-        this.setForkDisputed(Number(dispute.forkCnt));
-        return this.rechallengeRecursive(dispute);
+    private createDefaultDisputeStruct(): {
+        dispute: DisputeStruct;
+        timeout: TimeoutStruct;
+    } {
+        const timeout: TimeoutStruct = {
+            participant: ethers.ZeroAddress,
+            blockHeight: ethers.MaxUint256,
+            minTimeStamp: ethers.MaxUint256,
+            forkCnt: 0,
+            isForced: false,
+            previousBlockProducer: ethers.ZeroAddress,
+            previousBlockProducerPostedCalldata: false
+        };
+
+        let dispute: DisputeStruct = {
+            channelId: this.channelId,
+            genesisStateSnapshotHash: "",
+            latestStateSnapshotHash: "",
+            stateProof: {
+                forkProof: {
+                    forkMilestoneProofs: []
+                },
+                signedBlocks: []
+            },
+            fraudProofs: [],
+            onChainSlashes: [],
+            onChainLatestJoinChannelBlockHash: "",
+            outputStateSnapshotHash: "",
+            exitChannelBlocks: [],
+            disputeAuditingDataHash: "",
+            disputer: this.signerAddress,
+            disputeIndex: 0,
+            previousRecursiveDisputeIndex: ethers.MaxUint256,
+            timeout,
+            selfRemoval: false
+        };
+        return { dispute, timeout };
     }
 
-    //Creates a dispute based on the generated proofs or optimistically timeouts (folds) the provided participant
-    public async createDispute(
-        forkCnt: BigNumberish,
-        foldedParticipant: AddressLike,
-        foldedTransactionCnt: BigNumberish,
-        proofs: ProofStruct[]
-    ): Promise<void> {
-        if (foldedParticipant != NO_PARTICIPANT_TO_FOLD) {
-            console.log("DisputeHandler - createDispute - Timeout");
-        }
+    private async createDisputeStruct(
+        forkCnt: number,
+        transactionCnt: number,
+        proofs: ProofStruct[],
+        timestamp: BigNumberish,
+        timeout?: TimeoutStruct
+    ): Promise<{
+        dispute: DisputeStruct;
+        disputeAuditingData: DisputeAuditingDataStruct;
+    }> {
+        const disputeIndex =
+            await this.stateChannelManagerContract.getDisputeLength(
```

Comment:
forkCnt is disputeIndex no need to query the chain. In fact if you query the chain someone could have already created the dispute on-chain so the length increased so you would get a higher value than what you would expect and in that case you would get slashed. If you're disputing fork K the dispute index is K and if someone disputed K before you, on-chain your tx would fail and you wouldn't get slashed